zholobov.oleg/S1

OlegZholobov opened the pull request:
Fwd + Ring
Постараюсь сделать на 7 баллов
![image](https://github.com/user-attachments/assets/91225ed9-d31a-43d9-85bf-c48da4c09c5d)

OlegZholobov committed:
initial commit for S1

OlegZholobov committed:
added constructor and Node

OlegZholobov committed:
implemented CircularFwdListIterator

OlegZholobov committed:
implemented CircularFwdListConstIterator

OlegZholobov committed:
defined interface for CircularFwdList

OlegZholobov committed:
extended interface for CircularFwdList

OlegZholobov committed:
implemented some methods and basic tests

methods:
ctor, dtor, front/end, push_front

iterators:
use pointer to pointer to distinguish begin and end

OlegZholobov committed:
fixed push_front

OlegZholobov committed:
reworked and acceptance should pass now

take Node_t out of CircularFwdList class into separate file
number of fixes in CircularFwdList

OlegZholobov committed:
added some methods and fixed CircularFwdListConstIterator

added methods:
pop_*, swap, clear

TODO error handling in main()

OlegZholobov committed:
added exceptions handling and  cbegin, cend, empty methods

OlegZholobov committed:
believing in acceptance pass

OlegZholobov committed:
believing in acceptance pass (part 2)

OlegZholobov committed:
believing in acceptance pass (final)

OlegZholobov committed:
believing in acceptance pass (last final)

OlegZholobov committed:
believing in acceptance pass (very last final)

OlegZholobov committed:
continue implementing methods and many UTs

implemented:
constructor with initializer list
fill constructors
some fixes
all UT for basic interface and extended interface
Some tests fail since corresponding methods not implemented yet.

Volgarenok commented:
> Fwd + Ring Постараюсь сделать на 7 баллов

Там несколько 7-ок в таблице. Приложите табличку

Volgarenok commented:
@OlegZholobov Ну хотя бы контент должен проходить мне кажется...

OlegZholobov committed:
implemented splice_after methods and added tests

tests for:
constructors,
front,
back,
empty,
size,
(move)assign op,
splice_after

OlegZholobov committed:
cg

OlegZholobov committed:
made a few enhancements

OlegZholobov committed:
added stubs for "extened with iterators" methods

OlegZholobov committed:
made tests for "extended with iterators" methods

also:
added before_begin/end methods
reworked FwdListNode class
many minor fixes

OlegZholobov committed:
implemented almost all "extended with iterators" methods

Most tests now pass.
There are still a few errors in splice_after method resulting 9 tests to fail.
Will fix in next commit(s)

OlegZholobov committed:
cg

OlegZholobov committed:
fixed leak

OlegZholobov committed:
completed "extended with iterators" methods

implemented all needed methods
added some checks in unit tests
fixed errors per unit tests

OlegZholobov committed:
some optimizations

Volgarenok commented on zholobov.oleg/S1/main.cpp:
```diff
@@ -0,0 +1,129 @@
+    while (input.good()) {
+      std::string line;
+      std::getline(input, line);
+      std::stringstream ss(line);
```
А вы преамбулу к работе читали вообще?

Volgarenok commented on zholobov.oleg/S1/main.cpp:
```diff
@@ -0,0 +1,129 @@
+    std::ostream& out_;
+    std::string delimiter_;
+    bool first_;
+  };
```
Я не очень понимаю зачем это реализовывать. На практике мы рассмотрели это как пример, но только чтоб концепцию продемонстрировать. Тут у вас получается, что каждый раз лишняя проверка делается: выводить или не выводить разделитель

OlegZholobov replied:
С помощью этого SequencePrinter'а удобно выводить списки без лишних пробелов спереди или сзади не используя некрасивый способ вывода первого элемента, а затем в цикле всех остальных. А просто цикл по всем элементам, в том числе и range-based for.
Если говорить о проверке, нужно ли выводить разделитель при выводе каждого элемента, как получилось в этой реализации, то я немного переделал SequencePrinter, так что теперь там нет лишних проверок.

Volgarenok commented on zholobov.oleg/S1/main.cpp:
```diff
@@ -0,0 +1,129 @@
+      out << "0\n";
+    } else {
+      SequencePrinter< ElemType > printer(out);
+      for (const auto& elem : sum_list) {
```
CG: обратите внимание на требования к расстановке пробелов у двоеточия

Volgarenok commented on zholobov.oleg/S1/FwdListNode.hpp:
```diff
@@ -0,0 +1,42 @@
+    explicit FwdListNode(const T& val, FwdListNodeBase* next_node = nullptr):
+      FwdListNodeBase(next_node), value(val) {}
+    explicit FwdListNode(T&& val, FwdListNodeBase* next_node = nullptr):
+      FwdListNodeBase(next_node), value(std::move(val)) {}
```
CG просто мрак

Volgarenok commented on zholobov.oleg/S1/FwdListNode.hpp:
```diff
@@ -0,0 +1,42 @@
+  };
+
+  template < typename T >
+  struct FwdListNode: public FwdListNodeBase {
```
А зачем два класса делать?

OlegZholobov replied:
Тут идея в том, что FwdListNodeBase является базовым классом для класса представляющего узел односвязного списка FwdListNode (который только добавляет сами данные).
FwdListNodeBase служит только для манипулирования указателями, чтобы не обращать внимание на данные хранящиеся в узлах.
Идея была позаимствована из стандартного std::forward_list (у меня путь к нему /usr/include/c++/10/bits/forward_list.h). Там есть структура struct _Fwd_list_node_base выполняюжщая ту же роль. И struct _Fwd_list_node которая по назначению аналогична моему FwdListNode.
При помощи этого оказалось удобно хранить в самом контейнере указатель на первый элемент (FwdListNodeBase* head_), не храня при этом сами данные (то есть без Fake Node). Ведь приходится ещё хранить указатель FwdListNodeBase* before_head_ на этот head_, для простой реализации  before_begin и вставки элементов в начало списка.

Volgarenok commented on zholobov.oleg/S1/FwdListNode.hpp:
```diff
@@ -0,0 +1,42 @@
+  struct FwdListNode: public FwdListNodeBase {
+    T value;
+    FwdListNode():
+      FwdListNodeBase(), value() {}
```
Для чего такая нода может понадобится?

OlegZholobov replied:
см предыдущий ответ

Volgarenok replied:
Идея отличная (на мой взгляд нет) за исключением того, что накладывает дополнительные требования на конструктор по умолчанию для T

Volgarenok commented on zholobov.oleg/S1/CircularFwdListIterators.hpp:
```diff
@@ -0,0 +1,184 @@
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
```
Вы совсем что ли ку-ку?

OlegZholobov replied:
убрал

Volgarenok commented on zholobov.oleg/S1/CircularFwdListIterators.hpp:
```diff
@@ -0,0 +1,184 @@
+
+  private:
+    explicit CircularFwdListIterator(FwdListNodeBase** node);
+    FwdListNodeBase** node_;
```
А зачем тут лишняя косвенность?

OlegZholobov replied:
Поскольку список у меня Циклический, то если делать обычным образом, то для не пустого списка CircularFwdList::begin() равен CircularFwdList::end(). При этом становится невозможным пользоваться стандартным паттерном: for (auto it = list.begin(); it != list.end(); ++it) ...
range-based for loop очевидно тоже не будет работать.
А я хотел чтобы это всё работало.
Чтобы добиться того, чтобы begin() и end() различались, пришлось сделать так, чтобы итератор хранил указатель на указатель предыдущего элемента (который в моем случае FwdListNodeBase*).
Получается, что итератор возвращаемый методом begin() по сути содержит адрес указателя на первый узел: std::addressof(head_), а итератор, возвращаемый методом end() - содержит адрес указателя последнего узла списка т.е. std::addressof(tail_->next). И они очевидно разные. Код немного усложнился из-за того что теперь приходится лишний раз разыменовывания делать, но, проблема решена.

Volgarenok commented on zholobov.oleg/S1/test-list.cpp:
```diff
@@ -0,0 +1,516 @@
+    }
+  }
+  return out;
+}
```
Зачем глобально этот оператор определять?

OlegZholobov replied:
переделал

Volgarenok commented on zholobov.oleg/S1/test-list.cpp:
```diff
@@ -0,0 +1,516 @@
+  ss << list;
+  return ss.str();
+}
+
```
А просто функцию сделать без всяких перегрузок?

OlegZholobov replied:
объединил

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+  public:
+    using value_type = T;
+    using reference = value_type&;
+    using const_reference = const reference;
```
Не совсем правильно вы этот псевдоним сделали

OlegZholobov replied:
согласен, поправил

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+  } catch (const std::bad_alloc&) {
+    clear();
+    throw;
+  }
```
У вас делегированный конструктор вызывается. Следовательно деструктор будет вызываться при исключении. И его не нужно обрабатывать

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+    for (const auto& elem : other) {
+      push_back(elem);
+    }
+  } catch (const std::bad_alloc&) {
```
Ну строго говоря тут могут быть и другие исключения, которые генерируются типом T

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+void zholobov::CircularFwdList< T >::assign(std::initializer_list< T > init)
+{
+  CircularFwdList< T > temp;
+  for (const auto& elem : init) {
```
CG: обратите внимание на требования к расстановке пробелов у двоеточия

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+}
+
+template < typename T >
+void zholobov::CircularFwdList< T >::pop_back()
```
У однонаправленного списка методы работы с концом списка не очень эффективны. Добавлять эффективно можно и получать последний элемент можно (если хранить указаатель на хвост). Но не удалять

OlegZholobov replied:
убрал этот метод

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+void zholobov::CircularFwdList< T >::remove(const value_type& val)
+{
+  remove_if([&val](const T& value) { return value == val; });
+}
```
В CG есть требования по оформлению лямбда-выражений

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+
+template < typename T >
+void zholobov::CircularFwdList< T >::splice_after(const_iterator pos,
+                                                  CircularFwdList< T >& other, const_iterator it)
```
Переносы - мрак

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,628 @@
+{
+  FwdListNodeBase** pos_p = const_cast< FwdListNodeBase** >(pos.node_);
+  FwdListNodeBase** first_p = const_cast< FwdListNodeBase** >(first.node_);
+  FwdListNodeBase** last_p = const_cast< FwdListNodeBase** >(last.node_);
```
Я не понимаю зачем этой ерундой заниматься. Вам по кайфу касты что ли писать?

OlegZholobov committed:
changes per comments

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,607 @@
+
+template < typename T >
+void zholobov::CircularFwdList< T >::splice_after(
+    const_iterator pos, CircularFwdList< T >&& other, const_iterator first, const_iterator last)
```
Первый параметр остаётся на строке с именем функции, остальные - переносятся

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,607 @@
+}
+
+template < typename T >
+void zholobov::CircularFwdList< T >::splice_after(const_iterator pos, CircularFwdList< T >&& other)
```
Многова-то splice-ов

Volgarenok replied:
Нужно уменьшить дублирования

OlegZholobov replied:
Сделал шаблонные методы splice_after. Их количество уменьшилось в 2 раза.

Volgarenok commented on zholobov.oleg/S1/CircularFwdList.hpp:
```diff
@@ -0,0 +1,607 @@
+  } catch (...) {
+    clear();
+    throw;
+  }
```
Лишний обработчик

OlegZholobov committed:
changes per comments (part 2)

OlegZholobov committed:
Merge remote-tracking branch 'volgarenok/master' into zholobov.oleg/S1
